ffi/src/repr_c.rs

```diff
@@ -0,0 +1,690 @@
+    type FfiStore: Default;
+
+    /// Perform the conversion from [`Self`] into `[Self::ReprC]`
+    fn into_repr_c(self, store: &'itm mut Self::RustStore) -> Self::ReprC;
+
+    /// Perform the conversion from [`Self::ReprC`] into `[Self]`
+    ///
+    /// # Errors
+    ///
+    /// Check [`FfiReturn`]
+    ///
+    /// # Safety
+    ///
+    /// All conversions from a pointer must ensure pointer validity beforehand
+    unsafe fn try_from_repr_c(source: Self::ReprC, store: &'itm mut Self::FfiStore)
+        -> Result<Self>;
+}
+
+/// Provides the capability to use the type as an out-pointer.
+/// This trait mainly exists to be implemented on IR types.
+pub trait COutPtr: CType {
+    /// Out-pointer type
+    type OutPtr: OutPtrOf<Self::ReprC>;
+}
+
+/// Type that uses [`CType::Store`] during conversion
+// TODO: Clone requirement could be relaxed. For instance (OpaqueStruct1, OpaqueStruct2)
+// shouldn't require that OpaqueStruct1 and OpaqueStruct2 are Clone
+pub trait NonTransmute: Clone {}
+
+impl<T: NonTransmute + Clone> NonTransmute for &T {}
+impl<T: Ir + Clone> NonTransmute for IrBox<T, T::Type> {}
+impl<T: Ir + Clone> NonTransmute for IrSlice<'_, T, T::Type> {}
+impl<T: Ir + Clone> NonTransmute for IrVec<T, T::Type> {}
+
+impl<T: CType + Clone> CType for &T {
+    type ReprC = *const T::ReprC;
+}
+impl<'itm, T: CTypeConvert<'itm> + Clone> CTypeConvert<'itm> for &'itm T {
+    type RustStore = (Option<T::ReprC>, T::RustStore);
+    type FfiStore = (Option<T>, T::FfiStore);
+
+    fn into_repr_c(self, store: &'itm mut Self::RustStore) -> Self::ReprC {
+        store.0.insert(Clone::clone(self).into_repr_c(&mut store.1))
+    }
+
+    unsafe fn try_from_repr_c(
+        source: Self::ReprC,
+        store: &'itm mut Self::FfiStore,
+    ) -> Result<Self> {
+        Ok(store.0.insert(
+            CTypeConvert::try_from_repr_c(source.read(), &mut store.1)
+                .map(ManuallyDrop::<T>::new)
+                .map(|item| (*item).clone())?,
+        ))
+    }
+}
+
+// NOTE: `CType` cannot be implemented for `&mut T`
+impl<T: Ir<Type = U>, U: NonTransmute + CType + IrTypeOf<T>> CType for IrBox<T, U> {
+    type ReprC = *mut U::ReprC;
+}
+impl<'itm, T: Ir<Type = U> + 'itm, U: NonTransmute + CTypeConvert<'itm> + IrTypeOf<T>>
+    CTypeConvert<'itm> for IrBox<T, U>
+{
+    type RustStore = (Option<U::ReprC>, U::RustStore);
+    type FfiStore = U::FfiStore;
+
+    fn into_repr_c(self, store: &'itm mut Self::RustStore) -> Self::ReprC {
+        let unboxed: U = IrTypeOf::into_ir(*self.0);
+        store.0.insert(unboxed.into_repr_c(&mut store.1))
+    }
+    unsafe fn try_from_repr_c(
+        source: Self::ReprC,
+        store: &'itm mut Self::FfiStore,
+    ) -> Result<Self> {
+        if source.as_mut().is_some() {
+            return Err(FfiReturn::ArgIsNull);
+        }
+
+        U::try_from_repr_c(source.read(), store)
+            .map(ManuallyDrop::new)
+            .map(|item| (*item).clone())
```

Comment:
Because there is no ownership transfer:
1. a pointer `*mut T::Type::ReprC` is received
2. you read from the pointer to obtain `T::Type::ReprC`
3. you convert into an owned type with `try_from_repr_c` which gives you `T::Type`
4. `T::Type` is owned so you have to quickly remove the "ownership" by deciding to drop it manually (so that panic doesn't occur in the meantime)
5. Now you can clone it, which is all you wanted since you cannot take ownership over FFI

I can try explain with an example if this isn't clear